File: slp/backend-dotnet/Features/Source/SourceRepository.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using backend_dotnet.Data;

namespace backend_dotnet.Features.Source;

public class SourceRepository : ISourceRepository
{
    private readonly AppDbContext _context;

    public SourceRepository(AppDbContext context) => _context = context;

    public async Task<Source?> GetByIdAsync(int id) =>
        await _context.Sources.FirstOrDefaultAsync(s => s.Id == id && s.DeletedAt == null);

    public async Task<(IEnumerable<Source> Items, int Total)> GetUserSourcesAsync(
        int userId,
        SourceQueryParams query,
        bool includeDeleted = false)
    {
        var q = _context.Sources.Where(s => s.UserId == userId);

        if (!includeDeleted)
            q = q.Where(s => s.DeletedAt == null);

        // ── Filter: title search (case-insensitive) ─────────────────────────
        if (!string.IsNullOrWhiteSpace(query.Search))
        {
            var term = query.Search.Trim().ToLower();
            q = q.Where(s => s.Title.ToLower().Contains(term));
        }

        // ── Filter: type exact match ────────────────────────────────────────
        if (!string.IsNullOrWhiteSpace(query.Type))
        {
            var type = query.Type.Trim().ToLower();
            q = q.Where(s => s.Type == type);
        }

        var total = await q.CountAsync();

        var pageSize = Math.Clamp(query.PageSize, 1, 100);
        var page = Math.Max(query.Page, 1);

        var items = await q
            .OrderByDescending(s => s.CreatedAt)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return (items, total);
    }

    public async Task<Source> CreateAsync(Source source)
    {
        _context.Sources.Add(source);
        await _context.SaveChangesAsync();
        return source;
    }

    public async Task UpdateAsync(Source source)
    {
        source.UpdatedAt = DateTime.UtcNow;
        _context.Sources.Update(source);
        await _context.SaveChangesAsync();
    }

    public async Task SoftDeleteAsync(int id)
    {
        var source = await _context.Sources.FindAsync(id);
        if (source != null)
        {
            source.DeletedAt = DateTime.UtcNow;
            await _context.SaveChangesAsync();
        }
    }

    public async Task<bool> ExistsAsync(int id) =>
        await _context.Sources.AnyAsync(s => s.Id == id && s.DeletedAt == null);
}

File: slp/backend-nestjs/src/modules/source/source.repository.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, LessThanOrEqual, Like, MoreThanOrEqual, IsNull } from 'typeorm';
import { Source } from './source.entity';
import { SourceQueryParams } from './source.dto';

export interface ISourceRepository {
  getById(id: number): Promise<Source | null>;
  getUserSources(
    userId: number,
    query: SourceQueryParams,
    includeDeleted?: boolean,
  ): Promise<{ items: Source[]; total: number }>;
  create(source: Source): Promise<Source>;
  update(source: Source): Promise<void>;
  softDelete(id: number): Promise<void>;
  exists(id: number): Promise<boolean>;
}

@Injectable()
export class SourceRepository implements ISourceRepository {
  constructor(
    @InjectRepository(Source)
    private readonly repo: Repository<Source>,
  ) {}

  async getById(id: number): Promise<Source | null> {
    return this.repo.findOne({
      where: { id, deletedAt: IsNull() },
    });
  }

  async getUserSources(
    userId: number,
    query: SourceQueryParams,
    includeDeleted = false,
  ): Promise<{ items: Source[]; total: number }> {
    const qb = this.repo.createQueryBuilder('source')
      .where('source.userId = :userId', { userId });

    if (!includeDeleted) {
      qb.andWhere('source.deletedAt IS NULL');
    }

    if (query.search) {
      qb.andWhere('LOWER(source.title) LIKE LOWER(:search)', { search: `%${query.search}%` });
    }

    if (query.type) {
      qb.andWhere('source.type = :type', { type: query.type });
    }

    const total = await qb.getCount();

    const page = Math.max(query.page || 1, 1);
    const pageSize = Math.min(Math.max(query.pageSize || 20, 1), 100);
    const skip = (page - 1) * pageSize;

    const items = await qb
      .orderBy('source.createdAt', 'DESC')
      .skip(skip)
      .take(pageSize)
      .getMany();

    return { items, total };
  }

  async create(source: Source): Promise<Source> {
    return this.repo.save(source);
  }

  async update(source: Source): Promise<void> {
    source.updatedAt = new Date();
    await this.repo.save(source);
  }

  async softDelete(id: number): Promise<void> {
    await this.repo.update(id, { deletedAt: new Date() });
  }

  async exists(id: number): Promise<boolean> {
    const count = await this.repo.count({
      where: { id, deletedAt: IsNull() },
    });
    return count > 0;
  }
}